File: quant1x/contrib/data/tdx/protocol.py
# -*- coding: UTF-8 -*-
# Copyright (c) Quant1X <wangfengxy@sina.cn>.
# Licensed under the MIT License.

import abc
import struct
import zlib

from quant1x.net.conn import ConnectionHandle
from quant1x.log import logger
from .helpers import msg_sequence_id
from .command import FLAG_GENERIC, FLAG_UNCOMPRESSED
from .command import Command

class Stringable(abc.ABC):
    @abc.abstractmethod
    def to_string(self) -> str:
        raise NotImplementedError()


class Sizeable(abc.ABC):
    @abc.abstractmethod
    def byte_size(self) -> int:
        raise NotImplementedError()

class RequestHeader(Stringable, Sizeable, abc.ABC):
    zip_flag:    int # u8
    """压缩标识, u8"""
    sequence_id:  int # u32
    """消息ID, u32"""
    packet_type: int # u8
    """包类型, u8"""
    body_wire_len:    int # u16
    """包长度1, u16"""
    body_raw_len:    int # u16
    """包长度2, u16"""
    command:     Command # u16
    """命令字, u16"""

    def __init__(self, command: Command, flags: int = FLAG_UNCOMPRESSED):
        self.zip_flag = flags
        self.sequence_id = msg_sequence_id()
        self.packet_type = 0x01
        self.body_wire_len = 0
        self.body_raw_len = 0
        self.command = command
    
    def to_string(self) -> str:
        class_name = self.__class__.__name__
        return (
            f"{class_name}(zip_flag: {self.zip_flag}, sequence_id: {self.sequence_id}, "
            f"packet_type: {self.packet_type}, body_wire_len: {self.body_wire_len}, "
            f"body_raw_len: {self.body_raw_len}, command: {self.command})"
        )
    
    def byte_size(self) -> int:
        return 12  # 固定 12 字节
    
    def serialize(self) -> bytes:
        """
        将 header 字段按小端字节序打包为二进制数据
        
        Args:
            无显式参数，使用类实例属性
        
        Returns:
            bytes: 打包后的二进制数据，包含以下字段按顺序排列：
                - zip_flag (1字节)
                - sequence_id (4字节)
                - packet_type (1字节)
                - body_wire_len (2字节)
                - body_raw_len (2字节)
                - command (2字节)
        """
        return struct.pack(
            '<B I B H H H',
            self.zip_flag,
            self.sequence_id,
            self.packet_type,
            self.body_wire_len,
            self.body_raw_len,
            self.command.value & 0xFFFF
        )
    

class ResponseHeader(Stringable, Sizeable, abc.ABC):
    magic_number: int # u32
    """保留字段, u32"""
    zip_flag:   int # u8
    """压缩标识, u8"""
    sequence_id: int # u32
    """消息ID, u32"""
    packet_type:         int # u8
    """保留字段, u8"""
    command:    Command # u16
    """命令字, u16"""
    body_wire_len:   int # u16
    """压缩后大小, u16"""
    body_raw_len: int # u16
    """解压后/原始大小, u16"""
    
    def to_string(self) -> str:
        class_name = self.__class__.__name__
        return (
            f"{class_name}(magic_number: {self.magic_number}, zip_flag: {self.zip_flag}, sequence_id: {self.sequence_id}, "
            f"packet_type: {self.packet_type}, command: {self.command}, body_wire_len: {self.body_wire_len}, "
            f"body_raw_len: {self.body_raw_len})"
        )
    
    def byte_size(self) -> int:
        return 16
    
    def deserialize(self, data: bytes) -> None:
        """
        解析协议头信息
        
        Args:
            data: 协议头字节数据
        """
        # 解析协议头格式：I1(4字节), zip_flag(1字节), sequence_id(4字节), packet_type(1字节), command(2字节), body_wire_len(2字节), body_raw_len(2字节)
        self.magic_number, self.zip_flag, self.sequence_id, self.packet_type, cmd_value, self.body_wire_len, self.body_raw_len = struct.unpack('<I B I B H H H', data)
        # - packet_type (unsigned char): 1字节无符号整数
        # - command (unsigned short): 2字节无符号整数
        # - body_wire_len (unsigned short): 2字节无符号整数
        # - command (unsigned short): 2字节无符号整数
        # 将整数命令值转换为 Command 枚举
        
        try:
            self.command = Command(cmd_value)
        except ValueError:
            # 如果找不到对应的枚举，可以创建一个默认值或抛出异常
            logger.exception(f"警告: 未知的命令值 0x{cmd_value:04x}")
            # 或者设置一个默认值
            self.command = Command.UNKNOWN
        # - body_wire_len (unsigned short): 2字节无符号整数
        # - body_raw_len (unsigned short): 2字节无符号整数


class BaseMessage(abc.ABC):
    """
    消息基类
    
    用于处理消息头和消息体的解析和序列化。
    """
    def __init__(self, command: Command, flags: int = FLAG_UNCOMPRESSED):
        self.request_header = RequestHeader(command=command, flags=flags)
        self.response_header = ResponseHeader()
        self.reply = None
    
    @abc.abstractmethod
    def serialize_request_body(self) -> bytes:
        """
        将消息体序列化为二进制数据。

        Returns:
            bytes: 序列化后的二进制数据
        """
        raise NotImplementedError()
    
    def serialize_request(self) -> bytes:
        """
        将消息头和消息体序列化为二进制数据。
        
        Returns:
            bytes: 序列化后的二进制数据，包含以下内容：
                - 消息头(12字节)
                - 消息体(可变长度)
        """
        body_bytes = self.serialize_request_body()  # 安全：子类一定实现了
        self.request_header.body_wire_len = 2 + len(body_bytes)
        self.request_header.body_raw_len = 2 + len(body_bytes)
        return self.request_header.serialize() + body_bytes
    
    def deserialize_response_header(self, data: bytes) -> None:
        """
        从二进制数据中解析消息头。
        
        Args:
            data (bytes): 二进制数据
        """
        self.response_header.deserialize(data)
    
    @abc.abstractmethod
    def deserialize_response_body(self, data: bytes) -> None:
        """
        从二进制数据中解析消息体（抽象方法，需子类实现）。
        
        Args:
            data (bytes): 包含消息体的二进制数据
        
        Raises:
            NotImplementedError: 当子类未实现该方法时抛出
        """
        raise NotImplementedError()
    

def _recv_exact(conn_like: ConnectionHandle, n: int) -> bytes:
    """
    从支持 `recv(n)` 的对象读取恰好 `n` 字节的数据。
    
    Args:
        conn_like (ConnectionHandle): 实现 `recv` 方法的连接对象，用于屏蔽对原始 socket 的直接访问
        n (int): 需要读取的字节数
    
    Returns:
        bytes: 读取到的字节数据
    
    Raises:
        ConnectionError: 当连接在读取过程中关闭时抛出
    """
    buf = bytearray()
    while len(buf) < n:
        chunk = conn_like.recv(n - len(buf))
        if not chunk:
            raise ConnectionError("socket closed while reading")
        buf.extend(chunk)
    return bytes(buf)


def process_level1_new(conn_handle: ConnectionHandle, msg: BaseMessage) -> None:
    req_buf = msg.serialize_request()
    logger.debug(f"process_level1: request={msg.request_header.to_string()}")
    logger.debug(f"process_level1: req_buf={req_buf.hex()}")
    conn_handle.sendall(req_buf)

    # # 读取 16 字节响应头
    # hdr = _recv_exact(conn_handle, 16)
    #
    # # 解析头部: <I B I B H H H> => u32, u8, u32, u8, u16, u16, u16
    # i1, zip_flag, seq_id, packet_type, method, body_wire_len, body_raw_len = struct.unpack('<IBIBHHH', hdr)
    #
    # if body_wire_len == 0:
    #     return
    logger.debug(f"process_level1: response_header.byte_size={msg.response_header.byte_size()}")
    resp_header_bytes = _recv_exact(conn_handle, msg.response_header.byte_size())
    msg.response_header.deserialize(resp_header_bytes)
    if msg.response_header.body_wire_len == 0:
        return
    logger.debug(f"process_level1: response_header={msg.response_header.to_string()}")
    resp_body_bytes = _recv_exact(conn_handle, msg.response_header.body_wire_len)
    if msg.response_header.body_wire_len != msg.response_header.body_raw_len:
        # 如果压缩长度与解压长度不一致，则为 zlib 压缩数据，需要解压
        resp_body_bytes = zlib.decompress(resp_body_bytes)
    
    msg.deserialize_response_body(resp_body_bytes)
    #logger.debug(f"process_level1: response_body={msg.reply}")

from quant1x.net.handler import NetworkOperationHandler

class StandardProtocolHandler(NetworkOperationHandler):
    """标准协议处理器, 执行Synchronize1/Synchronize2握手和心跳。

    此实现调用`level1.protocol`来序列化请求, 并在提供的套接字上执行阻塞读/写.
    """

    def handshake(self, conn) -> bool:
        try:
            from .level1 import StdLogin
            #from .level1.ext import Synchronize as ext_Synchronize2

            msg1 = StdLogin()
            process_level1_new(conn, msg1)

            #msg2 = ext_Synchronize2()
            #process_level1_new(conn, msg2)
            return True
        except Exception as e:
            logger.exception('StandardProtocolHandler.handshake failed: {}', e)
            return False

    def keepalive(self, conn) -> bool:
        try:
            from .level1 import Heartbeat

            msg = Heartbeat()
            process_level1_new(conn, msg)
            return True
        except Exception as e:
            logger.exception('StandardProtocolHandler.keepalive failed: {}', e)
            return False

class ExtensionProtocolHandler(NetworkOperationHandler):
    """标准协议处理器, 执行ext_hello握手和心跳。

    此实现调用`level1.protocol`来序列化请求，并在提供的套接字上执行阻塞读/写。
    """

    def handshake(self, conn) -> bool:
        # 使用阻塞请求助手执行Synchronize1然后Synchronize2
        try:
            from .level1.ext import Synchronize
            
            req = Synchronize()
            process_level1_new(conn, req)
            return req.success
        except Exception as e:
            # 使用调试日志以避免在服务器检测期间产生噪音
            logger.exception('ExtensionProtocolHandler.handshake failed: {}', e)
            return False

    def keepalive(self, conn) -> bool:
        try:
            from .level1.ext import InstrumentCount
            
            req = InstrumentCount()
            process_level1_new(conn, req)
            return req.reply.get("count", 0) > 0
        except Exception as e:
            # 使用调试日志以避免在服务器检测期间产生噪音
            logger.exception('ExtensionProtocolHandler.keepalive failed: {}', e)
            return False

File: quant1x/contrib/data/tdx/protocol.rs
// -*- coding: utf-8 -*-
// Copyright (c) Quant1X <wangfengxy@sina.cn>.
// Licensed under the MIT License.
//
// TDX protocol — 协议头与消息基类
// 对应 Python quant1x/contrib/data/tdx/protocol.py

use crate::std::BinaryStream;
use flate2::read::ZlibDecoder;
use std::io::{Read, Write};

use super::command::{Command, FLAG_UNCOMPRESSED};
use super::helpers::msg_sequence_id;

// ============================================================
// 请求头 (12 字节)
// ============================================================

/// 请求头，对应 Python `RequestHeader`
///
/// 布局 (小端): zip_flag(u8) + sequence_id(u32) + packet_type(u8) + body_wire_len(u16) + body_raw_len(u16) + command(u16) = 12 字节
#[derive(Debug, Clone)]
pub struct RequestHeader {
    pub zip_flag: u8,
    pub sequence_id: u32,
    pub packet_type: u8,
    pub body_wire_len: u16,
    pub body_raw_len: u16,
    pub command: Command,
}

impl RequestHeader {
    /// 创建请求头，自动分配 sequence_id
    pub fn new(cmd: Command, flags: u8) -> Self {
        Self {
            zip_flag: flags,
            sequence_id: msg_sequence_id(),
            packet_type: 0x01,
            body_wire_len: 0,
            body_raw_len: 0,
            command: cmd,
        }
    }

    /// 固定 12 字节
    pub fn byte_size(&self) -> usize {
        12
    }

    /// 序列化为小端字节数组
    pub fn serialize(&self) -> Vec<u8> {
        let mut bs = BinaryStream::new();
        bs.push_u8(self.zip_flag);
        bs.push_u32(self.sequence_id);
        bs.push_u8(self.packet_type);
        bs.push_u16(self.body_wire_len);
        bs.push_u16(self.body_raw_len);
        bs.push_u16(self.command.value);
        bs.data().clone()
    }

    pub fn to_string(&self) -> String {
        format!(
            "RequestHeader(zip_flag: {}, sequence_id: {}, packet_type: {}, body_wire_len: {}, body_raw_len: {}, command: {:?})",
            self.zip_flag, self.sequence_id, self.packet_type, self.body_wire_len, self.body_raw_len, self.command
        )
    }
}

// ============================================================
// 响应头 (16 字节)
// ============================================================

/// 响应头，对应 Python `ResponseHeader`
///
/// 布局 (小端): magic_number(u32) + zip_flag(u8) + sequence_id(u32) + packet_type(u8) + command(u16) + body_wire_len(u16) + body_raw_len(u16) = 16 字节
#[derive(Debug, Clone)]
pub struct ResponseHeader {
    pub magic_number: u32,
    pub zip_flag: u8,
    pub sequence_id: u32,
    pub packet_type: u8,
    pub command: Command,
    pub body_wire_len: u16,
    pub body_raw_len: u16,
}

impl ResponseHeader {
    pub fn new() -> Self {
        Self {
            magic_number: 0,
            zip_flag: 0,
            sequence_id: 0,
            packet_type: 0,
            command: super::command::CMD_UNKNOWN,
            body_wire_len: 0,
            body_raw_len: 0,
        }
    }

    /// 固定 16 字节
    pub fn byte_size(&self) -> usize {
        16
    }

    /// 从 16 字节数据反序列化
    pub fn deserialize(&mut self, data: &[u8]) -> Result<(), crate::std::DeserializeError> {
        let mut bs = BinaryStream::from_vec(data.to_vec());
        self.magic_number = bs.get_u32()?;
        self.zip_flag = bs.get_u8()?;
        self.sequence_id = bs.get_u32()?;
        self.packet_type = bs.get_u8()?;
        let cmd_value = bs.get_u16()?;
        self.body_wire_len = bs.get_u16()?;
        self.body_raw_len = bs.get_u16()?;

        // 尝试查找已知命令，未知则用 UNKNOWN
        self.command = Command::from_value(cmd_value)
            .copied()
            .unwrap_or_else(|| {
                log::warn!("未知的命令值 0x{:04x}", cmd_value);
                super::command::CMD_UNKNOWN
            });

        Ok(())
    }

    pub fn to_string(&self) -> String {
        format!(
            "ResponseHeader(magic_number: {}, zip_flag: {}, sequence_id: {}, packet_type: {}, command: {:?}, body_wire_len: {}, body_raw_len: {})",
            self.magic_number, self.zip_flag, self.sequence_id, self.packet_type, self.command, self.body_wire_len, self.body_raw_len
        )
    }
}

// ============================================================
// 消息基类 — BaseMessage
// ============================================================

/// 消息基类，对应 Python `BaseMessage`
///
/// 包含请求头和响应头，子类需实现 `serialize_request_body` 和 `deserialize_response_body`。
pub trait BaseMessage {
    fn request_header(&self) -> &RequestHeader;
    fn request_header_mut(&mut self) -> &mut RequestHeader;
    fn response_header(&self) -> &ResponseHeader;
    fn response_header_mut(&mut self) -> &mut ResponseHeader;

    /// 序列化请求体
    fn serialize_request_body(&mut self) -> Vec<u8>;

    /// 反序列化响应体
    fn deserialize_response_body(&mut self, data: &[u8]) -> Result<(), crate::std::DeserializeError>;

    /// 完整序列化请求（头 + 体）
    fn serialize_request(&mut self) -> Vec<u8> {
        let body = self.serialize_request_body();
        let header = self.request_header_mut();
        header.body_wire_len = 2 + body.len() as u16;
        header.body_raw_len = header.body_wire_len;

        let mut buf = header.serialize();
        buf.extend_from_slice(&body);
        buf
    }

    /// 反序列化响应头
    fn deserialize_response_header(&mut self, data: &[u8]) -> Result<(), crate::std::DeserializeError> {
        self.response_header_mut().deserialize(data)
    }

    /// 获取响应压缩体大小
    fn response_zip_size(&self) -> usize {
        self.response_header().body_wire_len as usize
    }

    /// 获取响应解压后大小
    fn response_unzip_size(&self) -> usize {
        self.response_header().body_raw_len as usize
    }
}

// ============================================================
// 底层 I/O — recv_exact
// ============================================================

/// 从实现了 `Read` 的对象读取恰好 `n` 字节
///
/// 对应 Python `_recv_exact`
///
/// 兼容阻塞和非阻塞 socket：遇到 WouldBlock 时短暂等待后重试。
pub fn recv_exact<R: Read>(reader: &mut R, n: usize) -> std::io::Result<Vec<u8>> {
    let mut buf = vec![0u8; n];
    let mut offset = 0;
    while offset < n {
        match reader.read(&mut buf[offset..]) {
            Ok(0) => {
                return Err(std::io::Error::new(
                    std::io::ErrorKind::UnexpectedEof,
                    "connection closed",
                ));
            }
            Ok(nread) => {
                offset += nread;
            }
            Err(ref e) if e.kind() == std::io::ErrorKind::WouldBlock => {
                std::thread::sleep(std::time::Duration::from_millis(10));
                continue;
            }
            Err(e) => return Err(e),
        }
    }
    Ok(buf)
}

// ============================================================
// zlib 解压
// ============================================================

/// 解压 zlib 编码的数据
fn unzip(body: Vec<u8>, unzipped_size: usize) -> std::io::Result<Vec<u8>> {
    let mut decoder = ZlibDecoder::new(&body[..]);
    let mut out = Vec::with_capacity(unzipped_size);
    decoder.read_to_end(&mut out)?;
    Ok(out)
}

// ============================================================
// process_level1 — 泛型协议处理
// ============================================================

/// 泛型 process 函数，对应 Python `process_level1_new`
///
/// 发送请求、读取响应头、按需解压响应体、解析响应体。
pub fn process_level1<M: BaseMessage, R: Read>(
    reader: &mut R,
    writer: &mut dyn std::io::Write,
    msg: &mut M,
) -> Result<(), crate::std::DeserializeError> {
    process_level1_impl(reader, writer, msg)
}

/// 对单一 Read+Write stream 执行 process_level1
/// 使用阻塞 std::net::TcpStream，已设置读写超时
pub fn process_level1_stream<M: BaseMessage, T: Read + Write>(
    stream: &mut T,
    msg: &mut M,
) -> Result<(), crate::std::DeserializeError> {
    let cmd_value = msg.request_header().command.value;
    let cmd_desc = msg.request_header().command.desc;

    // 1. 发送请求
    let req_buf = msg.serialize_request();
    log::debug!("process_level1_stream: request={}", msg.request_header().to_string());
    log::debug!("process_level1_stream: req_buf hex={}", hex::encode(&req_buf));
    stream
        .write_all(&req_buf)
        .map_err(|e| crate::std::DeserializeError::Other(e.to_string()))?;

    // 2. 读取 16 字节响应头
    let resp_header_len = msg.response_header().byte_size();
    let resp_header_bytes = recv_exact(stream, resp_header_len)
        .map_err(|e| crate::std::DeserializeError::Other(e.to_string()))?;

    // 3. 反序列化响应头
    msg.deserialize_response_header(&resp_header_bytes)?;

    // 4. 如果没有 body，直接返回
    if msg.response_zip_size() == 0 {
        return Ok(());
    }

    log::debug!("process_level1_stream: response_header={}", msg.response_header().to_string());

    // 5. 读取响应体
    let resp_body_bytes = recv_exact(stream, msg.response_zip_size())
        .map_err(|e| crate::std::DeserializeError::Other(e.to_string()))?;

    // 6. 如果压缩长度 != 解压长度，zlib 解压
    let final_body = if msg.response_zip_size() != msg.response_unzip_size() {
        unzip(resp_body_bytes, msg.response_unzip_size())
            .map_err(|e| crate::std::DeserializeError::Other(e.to_string()))?
    } else {
        resp_body_bytes
    };

    // 7. 反序列化响应体
    msg.deserialize_response_body(&final_body).map_err(|e| {
        crate::std::DeserializeError::Other(format!(
            "response body deserialize error for {} (0x{:04x}): {}",
            cmd_desc, cmd_value, e
        ))
    })?;

    Ok(())
}

fn process_level1_impl<M: BaseMessage, R: Read>(
    reader: &mut R,
    writer: &mut dyn std::io::Write,
    msg: &mut M,
) -> Result<(), crate::std::DeserializeError> {
    let cmd_value = msg.request_header().command.value;
    let cmd_desc = msg.request_header().command.desc;

    // 1. 发送请求
    let req_buf = msg.serialize_request();
    log::debug!("process_level1: request={}", msg.request_header().to_string());
    log::debug!("process_level1: req_buf hex={}", hex::encode(&req_buf));
    writer
        .write_all(&req_buf)
        .map_err(|e| crate::std::DeserializeError::Other(e.to_string()))?;

    // 2. 读取 16 字节响应头
    let resp_header_len = msg.response_header().byte_size();
    log::debug!("process_level1: response_header.byte_size={}", resp_header_len);
    let resp_header_bytes = recv_exact(reader, resp_header_len)
        .map_err(|e| crate::std::DeserializeError::Other(e.to_string()))?;

    // 3. 反序列化响应头
    msg.deserialize_response_header(&resp_header_bytes)?;

    // 4. 如果没有 body，直接返回
    if msg.response_zip_size() == 0 {
        return Ok(());
    }

    log::debug!("process_level1: response_header={}", msg.response_header().to_string());

    // 5. 读取响应体
    let resp_body_bytes = recv_exact(reader, msg.response_zip_size())
        .map_err(|e| crate::std::DeserializeError::Other(e.to_string()))?;

    // 6. 如果压缩长度 != 解压长度，zlib 解压
    let final_body = if msg.response_zip_size() != msg.response_unzip_size() {
        unzip(resp_body_bytes, msg.response_unzip_size())
            .map_err(|e| crate::std::DeserializeError::Other(e.to_string()))?
    } else {
        resp_body_bytes
    };

    // 7. 反序列化响应体
    msg.deserialize_response_body(&final_body).map_err(|e| {
        crate::std::DeserializeError::Other(format!(
            "response body deserialize error for {} (0x{:04x}): {}",
            cmd_desc, cmd_value, e
        ))
    })?;

    Ok(())
}

// ============================================================
// 测试
// ============================================================

// ============================================================
// StandardProtocolHandler — 标准行情连接处理器
// 对应 Python protocol.py StandardProtocolHandler
// ============================================================

use std::net::TcpStream as StdTcpStream;
use std::time::Duration;

use crate::io::operation_handler::NetworkOperationHandler;

use super::level1::std::hello::{Hello1Request, Hello2Request};
use super::level1::std::heartbeat::HeartbeatRequest;

pub struct StandardProtocolHandler;

impl NetworkOperationHandler for StandardProtocolHandler {
    fn handshake(&self, stream: &mut StdTcpStream) -> std::io::Result<()> {
        // Hello1
        let mut req1 = Hello1Request::new();
        match process_level1_stream(stream, &mut req1) {
            Ok(()) => {
                if req1.info.trim().is_empty() {
                    log::error!("StandardProtocolHandler::handshake Hello1 validation failed: empty info");
                    return Err(std::io::Error::new(std::io::ErrorKind::Other, "Hello1 response invalid or empty"));
                }
            }
            Err(e) => {
                log::error!("StandardProtocolHandler::handshake Hello1 failed: {}", e);
                return Err(std::io::Error::new(std::io::ErrorKind::Other, e.to_string()));
            }
        }

        // Hello2
        let mut req2 = Hello2Request::new();
        match process_level1_stream(stream, &mut req2) {
            Ok(()) => {
                if req2.info.trim().is_empty() {
                    log::error!("StandardProtocolHandler::handshake Hello2 validation failed: empty info");
                    return Err(std::io::Error::new(std::io::ErrorKind::Other, "Hello2 response invalid or empty"));
                }
            }
            Err(e) => {
                log::error!("StandardProtocolHandler::handshake Hello2 failed: {}", e);
                return Err(std::io::Error::new(std::io::ErrorKind::Other, e.to_string()));
            }
        }

        Ok(())
    }

    fn keepalive(&self, stream: &mut mio::net::TcpStream) -> std::io::Result<bool> {
        let mut req = HeartbeatRequest::new();
        match process_level1_stream(stream, &mut req) {
            Ok(()) => Ok(true),
            Err(_) => Ok(false),
        }
    }

    fn timeout(&self) -> Duration {
        Duration::from_secs(10)
    }

    fn check_interval(&self) -> Duration {
        Duration::from_secs(5)
    }
}

// ============================================================
// ExtensionProtocolHandler — 扩展行情连接处理器
// 对应 Python protocol.py ExtensionProtocolHandler
// ============================================================

use super::level1::ext::{ExtSynchronizeRequest, InstrumentCountRequest};

pub struct ExtensionProtocolHandler;

impl NetworkOperationHandler for ExtensionProtocolHandler {
    fn handshake(&self, stream: &mut StdTcpStream) -> std::io::Result<()> {
        let mut req = ExtSynchronizeRequest::new();
        match process_level1_stream(stream, &mut req) {
            Ok(()) if req.success => Ok(()),
            Ok(()) => Err(std::io::Error::new(
                std::io::ErrorKind::Other,
                "ExtensionProtocolHandler: synchronize failed (success=false)",
            )),
            Err(e) => Err(std::io::Error::new(std::io::ErrorKind::Other, e.to_string())),
        }
    }

    fn keepalive(&self, stream: &mut mio::net::TcpStream) -> std::io::Result<bool> {
        let mut req = InstrumentCountRequest::new();
        match process_level1_stream(stream, &mut req) {
            Ok(()) => Ok(req.count > 0),
            Err(_) => Ok(false),
        }
    }

    fn timeout(&self) -> Duration {
        Duration::from_secs(10)
    }

    fn check_interval(&self) -> Duration {
        Duration::from_secs(5)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use super::super::command::{STD_HEARTBEAT, FLAG_UNCOMPRESSED};

    #[test]
    fn test_request_header_byte_size() {
        let hdr = RequestHeader::new(STD_HEARTBEAT, FLAG_UNCOMPRESSED);
        assert_eq!(hdr.byte_size(), 12);
    }

    #[test]
    fn test_request_header_serialize() {
        let hdr = RequestHeader::new(STD_HEARTBEAT, FLAG_UNCOMPRESSED);
        let data = hdr.serialize();
        assert_eq!(data.len(), 12);
        // zip_flag
        assert_eq!(data[0], FLAG_UNCOMPRESSED);
        // packet_type
        assert_eq!(data[5], 0x01);
        // command value (little-endian)
        let cmd_bytes = &data[10..12];
        assert_eq!(u16::from_le_bytes([cmd_bytes[0], cmd_bytes[1]]), STD_HEARTBEAT.value);
    }

    #[test]
    fn test_response_header_deserialize() {
        // 构造一个简单的响应头: magic=0x11223344, zip_flag=0x0C, seq_id=42, pkt_type=0, cmd=0x0004, wire=100, raw=200
        let mut buf = Vec::new();
        buf.extend_from_slice(&0x11223344u32.to_le_bytes());  // magic_number
        buf.push(0x0C);                                        // zip_flag
        buf.extend_from_slice(&42u32.to_le_bytes());           // sequence_id
        buf.push(0x00);                                        // packet_type
        buf.extend_from_slice(&0x0004u16.to_le_bytes());       // command = STD_HEARTBEAT
        buf.extend_from_slice(&100u16.to_le_bytes());          // body_wire_len
        buf.extend_from_slice(&200u16.to_le_bytes());          // body_raw_len

        let mut hdr = ResponseHeader::new();
        hdr.deserialize(&buf).unwrap();

        assert_eq!(hdr.magic_number, 0x11223344);
        assert_eq!(hdr.zip_flag, 0x0C);
        assert_eq!(hdr.sequence_id, 42);
        assert_eq!(hdr.command.value, 0x0004);
        assert_eq!(hdr.body_wire_len, 100);
        assert_eq!(hdr.body_raw_len, 200);
    }

    #[test]
    fn test_response_header_unknown_command() {
        let mut buf = Vec::new();
        buf.extend_from_slice(&0u32.to_le_bytes());
        buf.push(0);
        buf.extend_from_slice(&0u32.to_le_bytes());
        buf.push(0);
        buf.extend_from_slice(&0xFFFFu16.to_le_bytes());  // unknown command
        buf.extend_from_slice(&0u16.to_le_bytes());
        buf.extend_from_slice(&0u16.to_le_bytes());

        let mut hdr = ResponseHeader::new();
        hdr.deserialize(&buf).unwrap();
        assert_eq!(hdr.command.value, super::super::command::CMD_UNKNOWN.value);
    }
}
